### DISCUSSION TITLE
/articles/sharepoint-framework-secure-spfx-solutions-post-domain-isolated-webpart-retirement/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#316] [Comments]
Learn how to secure SPFx solutions after Microsoft retires domain-isolated web parts and explore options for hardening your SPFx applications.
https://www.andrewconnell.com/articles/sharepoint-framework-secure-spfx-solutions-post-domain-isolated-webpart-retirement/

**Comment by jbrown-lucid** [comment]
Can you help me understand how validating the `appid` claim helps secure a Microsoft Entra ID-secured endpoint?  My understanding is that the SharePoint Online Web Client Extensibility app is used by _all_ SPFx webparts.  If a Sharepoint tenant has two webparts installed, one of which _should_ be able to call my Microsoft Entra ID-secured endpoint, and one of which _should not_ be able to call it, is there any way to prevent the unauthorized webpart from calling the endpoint without isolated webparts?

**Reply by andrewconnell** [reply]
@jbrown-lucid Said:
> ... is there any way to prevent the unauthorized webpart from calling the endpoint without isolated webparts

Nope.

### TL;DR

With isolated web parts, there was a way because the call could be tied to a specific domain. But now that isolated web parts are on the deprecation & retirement block, there's no way to do it.

There's no way for SharePoint Online to know what made the request for the access token because this can only be controlled using the host domain of the calling script. Without isolated web parts & their unique domains, now everything runs on the tenant. 

**So... there's no supported way to tie a request to a specific SPFx component.**

In fact, there's nothing stopping you from not using an SPFx component at all to get an access token. If you know the SPO API endpoint SPFx uses to request an access token, you can use the browser's dev tools to go into the console tab & submit a fetch request for an access token to an endpoint, like Microsoft Graph. 

SPO will authenticate with Entra ID using SPO's own Entra ID app registration and request an access token on behalf of the current user for Microsoft Graph. The access token SPO gets back and thus returned to the caller, will have all Microsoft Graph scopes granted to that SPO tenant.

**Reply by andrewconnell** [reply]
Furthermore, there's no link between the permissions defined in an SPFx component's **package-solution.json** file and the SPO tenant. Permission requests & grants (ie: everything in the SPO Admin Center's API Access page) have zero association with SPFx components.

That was the basis for my guidance in this article: **[Beware of Declarative Permissions in SharePoint Framework Projects](https://www.voitanos.io/blog/consider-avoiding-declarative-permissions-with-azure-ad-services-in-sharepoint-framework-projects/)**

> Developers can declare permissions in SharePoint Framework projects, but it isn't a good idea. In this article I explain why and suggest a better approach.
